[CIR] Add nothrow attribute to the call operation

Lancern opened the pull request:
This patch adds extra function attributes to the `cir.call` operation. The extra attributes now may contain a single `cir.nothrow` attribute that indicates whether the callee throws.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -368,4 +368,34 @@ def CIR_VisibilityAttr : CIR_Attr<"Visibility", "visibility"> {
+// ExtraFuncAttributesAttr
+//===----------------------------------------------------------------------===//
+
+def ExtraFuncAttributesAttr : CIR_Attr<"ExtraFuncAttributes", "extra"> {
```
@xlauko and I start discussing about whether we really need `ExtraFuncAttributesAttr`. Maybe this is a good point in time to decide on it, and perhaps we don't need to introduce this upstream. The original idea was to use this to share attributes that later in LLVM could be placed in both functions and calls, but currently there isn't anything special about it (IIRC) and LLVMDialect uses a more natural approach to represent these, seems like we could do similar?

andykaylor replied:
+1 for not upstreaming this if it isn't needed.

Lancern replied:
LGTM to not upstreaming this if it's unnecessary, but then should we still upstreaming the `#cir.nothrow` attribute? Now `#cir.nothrow` is defined as a `CIRUnitAttr`, but maybe we could just use a simple `UnitAttr` for it instead?

bcardosolopes replied:
Good catch, I think `CIRUnitAttr` is in a similar boat!

xlauko replied:
Yeah, I am trying to get it work without custom unit attributes and `ExtraFuncAttributesAttr` but it will take me a bit more time. I first need to finish assembly format for FuncOp, so its easier to pull out these extra attrs. But I am almost there.

bcardosolopes reviewed (commented):
Overall looks good, added a higher level comment about `ExtraFuncAttributesAttr`

andykaylor commented on clang/test/CIR/CodeGen/call.cpp:
```diff
@@ -102,3 +104,16 @@ void f12() {
+// CIR:       }
+
+// LLVM-LABEL: define void @_Z3f14v()
+// LLVM:         call void @_Z3f13v()
```
Is this capturing the complete LLVM IR. It appears that the incubator doesn't set a callsite attribute here, but classic codegen does. That would seem to indicate that we missed something in lowering.

Lancern replied:
Updated.

Lancern committed:
[CIR] Add nothrow attribute to call operations

This patch adds nothrow attribute to call operations to indicate the callee will
not throw.

This patch also contains a minor refactor. The construction of call operations
is now split into two steps. The CallOp::build function and the create function
in CIR builder now only accepts basic parameters including the callee and the
arguments. Other extra arguments, including any additional attributes, are set
later on the constructed op in a separate step.

Lancern commented:
I conducted a minor refactor in the latest commit.

Prior to the latest commit, CallOp build functions (namely, `CallOp::build`, `CIRBaseBuilderTy::createCallOp`, and `CIRBaseBuilderTy::createIndirectCallOp`) takes quite a lot parameters, including the extra optional call attributes. I'm worried that this could eventually explode the parameter list of build functions, making them difficult to read and use. The latest commit removes these extra parameters from the build functions, and instead set these attributes on the constructed call op later in a separate step.

bcardosolopes commented on clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h:
```diff
@@ -227,26 +227,24 @@ class CIRBaseBuilderTy : public mlir::OpBuilder {
 
   cir::CallOp createCallOp(mlir::Location loc, mlir::SymbolRefAttr callee,
-                           mlir::Type returnType, mlir::ValueRange operands,
-                           cir::SideEffect sideEffect = cir::SideEffect::All) {
```
Have you thought about just having a `ArrayRef<Attribute> attrs = {}`? This would allow the caller to already pass stuff in if available.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -611,6 +611,9 @@ static mlir::ParseResult parseCallCommon(mlir::OpAsmParser &parser,
     return mlir::failure();
 
+  if (parser.parseOptionalKeyword("nothrow").succeeded())
+    result.addAttribute("nothrow", mlir::UnitAttr::get(parser.getContext()));
```
Same for the literal string here.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -662,13 +665,17 @@ static void printCallCommon(mlir::Operation *op,
   printer << "(" << ops << ")";
 
+  if (isNothrow)
+    printer << " nothrow";
```
and here!

bcardosolopes reviewed (commented):
Thanks for the changes, more comments!

bcardosolopes reviewed (approved):
LGTM after addressing the remaining nits

bcardosolopes reviewed (commented):
Thanks for updating the other attributes names as well, way to go.
